File: 2021/06/25/main.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        i, n = 0, len(edges)
        p = [i for i in range(n+1)]
        def ifind(i):
            while i != p[i]:
                i = p[i]
            return i
        for f, t in edges:
            ff = ifind(f)
            ft = ifind(t)
            if ff == ft:
                break
            p[ff] = ft
            i += 1
        return edges[i]

File: 2021/06/25/main.go
package main

func findRedundantConnection(edges [][]int) []int {
	n := len(edges)
	p := make([]int, n+1)
	for i := 0; i < n; i++ {
		p[i] = i
	}
	var ifind func(int) int
	ifind = func(i int) int {
		for i != p[i] {
			i = p[i]
		}
		return i
	}

	var i int
	var e []int
	for i, e = range edges {
		f, t := e[0], e[1]
		pf := ifind(f)
		pt := ifind(t)
		if pf == pt {
			break
		}
		p[pf] = pt
	}
	return edges[i]
}
